Make commenting contextual

ericyd opened the pull request:
Per #23, the commenting in this extension does not work correctly.  Removing the line comment property from the language configuration fixes this issue and allows the standard `ctrl+/` key binding to work correctly for CFML files.

This would close #18 and addresses concerns raised in #15.

ericyd committed:
Remove line comment from language configuration

KamasamaK commented:
I don't think this would close #18 since the discussion became a request for commenting to be contextual, not to just always use the tag syntax. PR #22 changed the comment from the script to tag syntax, but the same complaint can still be levied now for when you are in the script context.

ericyd commented:
That's a good point, I didn't read #18 clearly.

I guess this wouldn't exactly close anything, but I actually get an error when I try to comment with `ctrl+/`, which is why I made this PR.
![image](https://user-images.githubusercontent.com/8379268/29892727-9b847e84-8d84-11e7-87f1-bb882a1f11a9.png)

I have updated my local version to fix this for myself, so if this is a specific issue on my machine then this probably isn't a worthwhile PR.

KamasamaK commented:
It's not just you. As I said in #23 this change is correct if the default is supposed to be tag syntax. Although it seems strange considering the file name is `cfscript.configuration.json`. Also, I would wager that modern CFML development is done more in script than tags, but someone is going to be unhappy with either default which is why #18 is a good idea.

ericyd committed:
add extension to provide contextual commenting

ericyd commented:
This is a WIP - I think it will be possible to incorporate some quasi-hacky contextual comment-switching with regexs, but I have to go to work now.

ericyd committed:
Update regular expressions to accurately determine code context

ericyd commented:
OK, I believe this is complete.

The updates proposed here would add a basic extension that shares a keybinding with the default line-comment keybinding, `ctrl+/`. This extension's keybinding would only activate when the document language ID is "lang-cfml".

When a line is commented with `ctrl+/`, the extension will try to determine the context of the line(s) being commented, and use the correct form of comments for the context.

Performance is a potential concern, and I do not have any benchmarks to lean on.  Anecdotally, I have used this on 6,000-line tag-based files and there was no perceivable performance degradation. However, this is obviously somewhat hardware- and file-dependent so this is still a consideration.

Below are several schematic examples of situations that were tested manually with this update. I am not familiar with writing tests for vscode that would prove it's performance, but if any other contributors know how to write effective tests for extensions that would be welcome.

**Pass**: Use cfscript comments in a cfscript file (no \<cf\> tags present)
```
component {
  // this.name = "commented"
  this.name = "right"
}
```

**Pass**: Use cfscript comments in a tag-based file within a \<cfscript\> tag
```
<cfoutput>something</cfoutput>
<cfscript>
  // line = "commented"
  line = "uncommented"
</cfscript>
```

**Pass**: Use tag comments in a tag-based file, except within a \<cfscript\> tag
```
<cfoutput>
  <!--- #myOldVar# --->
  #myNewVar#
</cfoutput>
<!--- <cfscript>
  no longer needed
  myOldVar = 42
</cfscript> --->
```

**Pass**: Use tag comments on single-line cfscript declarations
```
<!--- <cfscript>notNecessary = true</cfscript> --->
```

**Fail**: Use script comments within \<script\> tag
```
<script>
  <!--- var el = document.getElementById('el'); --->
</script>
```

**Fail**: Use tag comments on code that lies within a commented-out \<cfscript\> tag
```
<!--- <cfscript>   --->
  // this is a test
<!--- </cfscript> --->
```

Welcome additions to this PR:
* Discussion on additional necessary test cases
* Ideas on how to fix the failing test cases
* Any automated tests

KamasamaK commented:
This is great work so far.

[`LanguageConfiguration`](https://code.visualstudio.com/docs/extensionAPI/vscode-api#LanguageConfiguration) does not support the `autoClosingPairs` or `surroundingPairs` properties, so they should be removed. From my testing, you can also safely remove `brackets` as its absence does not appear to remove what is defined in the static language configuration.

You should hijack the block comment keybinding as well (`Shift+Alt+A`). This allows commenting only part of a line. Also, since you're not, it stays configured to the style of the last line comment you used.

I will try to do more testing this weekend.

ericyd commented:
Ah, good point. I almost never use block comments, but that is an important use-case and shouldn't be too hard to integrate.

I will not be able to contribute anything on this until next Tuesday (9/5) at the earliest so this may be a bit stagnant for a few days.

ericyd committed:
add block comment command and mocha dependency

ericyd committed:
refactor matcher functions and add tests

ericyd committed:
add block comment command

ericyd commented:
I added block comment support and refactored the extension a bit so I could write some basic tests for the context matching.

The tests should serve as documentation for the scenarios in which this will work and fail.  I was not able to successfully write tests that included the vscode API so I limited my tests to just the context matching logic.

One issue I noticed is that Windows and Linux do not share a keybinding for block comments, and I don't believe that the [`contributes.keybinding`](https://code.visualstudio.com/docs/extensionAPI/extension-points#_contributeskeybindings) property allows for specifying a Linux keybinding - I believe it assumes one keybinding for Windows and Linux with an optional Mac property.  The keybinding still works on Linux, but isn't shared with the default.  This should be a relatively minor issue as it is quite easy to adjust keybindings in vscode.

ericyd commented:
Found a bug I need to address: if you have a file that is tag-based and you go to the first line, above any tags, and try to add a line comment, it will use cfscript style comments.  Should be tag style in this case.
![image](https://user-images.githubusercontent.com/8379268/30232748-907921b0-94a6-11e7-9482-c747646fd2ae.png)

ericyd committed:
Make tag-based files use tag comments when commenting first line

ericyd commented:
Two updates:
1. Fixed bug noted above. Now, tag-based files will use tag comments when the comment is on the first line of the file
2. Added a try/catch around one of the functions that sometimes threw an error which prevented comments from inserting

There are probably edge cases out there that are not accounted for, but I think this update would greatly improve the current usability of this extension. Any concerns or discussion is welcome.

ilich commented:
Thank you very much your your help! Added comments support as the part of #18.

ericyd commented:
Are you planning to add this eventually? I noticed you added it to the changelog, but your comment indicates that you are finished considering updates to comments.

This PR does something slightly different than 18, though I understand this project is not your top priority right now.

ilich commented:
Right now I am about to start large project in CFML, so the extension's priority is higher. I am planning 1.0.0 release soon. I have just submitted new comments engine. The changelist is in #18. Please take a look and feel free to add the bits I am missing there. I tried the test cases from this PR. Thank you very much for them!

I am going to add dark theme with CFML tags highlighting and new snippets, and then publish new version to marketplace. Other things like code navigation is a bit more complex, and I am working on the language server at the moment.
